naptime-graphql/src/main/scala/org/coursera/naptime/ari/graphql/GraphqlSchemaProvider.scala

```diff
@@ -0,0 +1,105 @@
+package org.coursera.naptime.ari.graphql
+
+import javax.inject.Inject
+
+import com.linkedin.data.schema.DataSchema
+import com.linkedin.data.schema.RecordDataSchema
+import com.typesafe.scalalogging.StrictLogging
+import org.coursera.naptime.ari.SchemaProvider
+import org.coursera.naptime.schema.Resource
+import sangria.schema.ObjectType
+import sangria.schema.Schema
+
+import scala.util.control.NonFatal
+
+/**
+ * Provides GraphQL schemas for other components of the ARI+GraphQL system.
+ */
+trait GraphqlSchemaProvider {
+  def schema: Schema[org.coursera.naptime.ari.graphql.SangriaGraphQlContext, Any]
+}
+
+/**
+ * A GraphQL Schema Provider implementation.
+ *
+ * We compute the schema and cache it for performance reasons.
+ *
+ * Note: we assume that the schemaProvider can return different schemas over time. We also assume
+ * that they change relatively slowly. As a result, we do not bother taking a lock between querying
+ * the types and resourceSchemas collections. In case there is a change in collections between, we
+ * retry a fixed number of times, before giving up and operating with an inconsistent snapshot.
+ *
+ * Note: we rely on object identity to ensure an efficient set comparison. (This is a reasonably good
+ * approach, because we assume immutable collections. Therefore we know we will never skip re-computing
+ * when we should. If the schema provider returns sets that are not object identities, but are equal, then
+ * we unnecessarily re-compute. We feel this is reasonable, because if the SchemaProvider is not re-using
+ * the same sets, then that is generating unnecessary garbage, and that should be fixed.)
+ *
+ * @param schemaProvider A schema provider.
+ */
+class DefaultGraphqlSchemaProvider @Inject() (schemaProvider: SchemaProvider)
+  extends GraphqlSchemaProvider with StrictLogging {
+  private[this] var resourceSchemas: Set[Resource] = Set.empty
+  private[this] var typesSet: Set[DataSchema] = Set.empty
+  private[this] var cachedSchema: Schema[SangriaGraphQlContext, Any] = DefaultGraphqlSchemaProvider.EMPTY_SCHEMA
+
+  private[this] def consistentSnapshot(numRetries: Int = 2): (Set[Resource], Set[DataSchema]) = {
+    if (numRetries == 0) {
+      logger.warn(s"Could not compute a consistent snapshot of the schema provider.")
+      (schemaProvider.resourceSchemas, schemaProvider.types)
+    } else {
+      val resourceSchemas = schemaProvider.resourceSchemas
+      val types = schemaProvider.types
+      if ((schemaProvider.resourceSchemas eq resourceSchemas) && (schemaProvider.types eq types)) {
+        (resourceSchemas, types)
+      } else {
+        consistentSnapshot(numRetries - 1)
```

Comment:
do we need to set any timeout here, or are you just assuming that the time to consistency will be so small that simply running it again will result in a consistent view? this seems a little funny (and possibly over-engineering.)

is the concern that we might set resourceSchemas, then read resourceSchemas and typesSet, and then set typesSet? If so, maybe we just store them all together in a case class or something that'll reduce code complexity, even if its a bit less performant.

let's chat more tomorrow